### PR TITLE
Parallel

### PR DESCRIPTION
**Pull request by casey**
Version of #1562 using rust's native scoped threads.

### CODE REVIEW

File: src/justfile.rs [hunk 1]
```diff
@@ -253,9 +254,32 @@ impl<'src> Justfile<'src> {
...
+          if let Err(err) = Self::run_recipe(&context, recipe, arguments, &dotenv, search, &ran) {
+            let _ = tx.send(err);
+            true
+          } else {
+            false
+          }
+        };
+        if config.parallel {
+          scope.spawn(run);
+        } else if run() {
```
Comment (syphar) [inline src/justfile.rs:273]: nit: I found it confusing when reading the code to expect `true` for the error case, but that might just be me.

File: src/justfile.rs [hunk 2]
```diff
@@ -253,9 +254,32 @@ impl<'src> Justfile<'src> {
...
+          break;
```
Comment (syphar) [inline src/justfile.rs:274]: If what you wanted to achieve is an early exit when the first task fails, this won't work in this implementation. 

A `thread::scope` (also `crossbeam::scope`) will wait for all spawned threads to finish before exiting the scope. 

I also tested this manually.

File: src/justfile.rs [hunk 3]
```diff
@@ -311,46 +335,111 @@ impl<'src> Justfile<'src> {
...
+            true
+          } else {
+            false
+          }
+        };
+
+        if parallel {
+          scope.spawn(run);
+        } else if run() {
+          break;
```
Comment (syphar) [inline src/justfile.rs:376]: see above, the early exit won't work like this.

**Review by syphar — COMMENTED** [review]
apart from style I have one remark, coming from what you said about early exit when one of the tasks fails.

**Comment by casey** [comment]
@syphar Ah, good point about early termination. We definitely want that, since presumably, if users are using parallel execution, they have some long-running recipes and want them to terminate early. Hmm, that's annoying. There's no way to terminate a thread from outside of a thread in Rust, so the only way we could do this is by having a channel, and then when spawning commands doing something like:

```rust
loop {
  match rx.recv_timeout(Duration::from_millis(10)) {
    Err(RecvTimeoutErr::Timeout) => {}
    Err(RecvTimout::Disconnected) => { child.kill(); return Ok(None); }
    Ok(()) => { child.kill(); return Ok(None); }
  }
  match child.try_wait() {
    Ok(None) => {},
    Ok(Some(status)) => return Ok(Some(status)),
    Err(err) => return Err(err),
  }
}
```

**Comment by casey** [comment]
Other issues, which I haven't thought about:

- I don't know if this handles dependencies correctly. Previously, everything was synchronous, so if an invocation wasn't in `ran`, you would dispatch the dependencies, which would add themselves to ran. However, now, the dependencies might already be running, so you could double-dispatch them.
- Recipe output and line echoing will happen in random order. Which is fine. However, they might be intermingled within a single line, which would be illegible.

**Comment by syphar** [comment]
> @syphar Ah, good point about early termination. We definitely want that, since presumably, if users are using parallel execution, they have some long-running recipes and want them to terminate early. Hmm, that's annoying. There's no way to terminate a thread from outside of a thread in Rust, so the only way we could do this is by having a channel, and then when spawning commands doing something like:
> 
> 
> 
> ```rust
> 
> loop {
> 
>   match rx.recv_timeout(Duration::from_millis(10)) {
> 
>     Err(RecvTimeoutErr::Timeout) => {}
> 
>     Err(RecvTimout::Disconnected) => { child.kill(); return Ok(None); }
> 
>     Ok(()) => { child.kill(); return Ok(None); }
> 
>   }
> 
>   match child.try_wait() {
> 
>     Ok(None) => {},
> 
>     Ok(Some(status)) => return Ok(Some(status)),
> 
>     Err(err) => return Err(err),
> 
>   }
> 
> }
> 
> ```
> 
> 

Yeah, I imagine that would be a valid approach, killing threads is not really nice, but in case of `just`, killing the spawned process (or exiting between calling processes in the thread) is probably good enough.

**Comment by syphar** [comment]
> 
> - I don't know if this handles dependencies correctly. Previously, everything was synchronous, so if an invocation wasn't in `ran`, you would dispatch the dependencies, which would add themselves to ran. However, now, the dependencies might already be running, so you could double-dispatch them.


From what I remember when testing dependencies, the parallel threads where handled as a group, so all dependencies of these were handled before handling the first of them. And everything depending on any one of the parallel tasks were waiting for the whole group to finish. I might misremember this, definitely needs testing. 


> 
> - Recipe output and line echoing will happen in random order. Which is fine. However, they might be intermingled within a single line, which would be illegible.

I don't know enough about stdout buffering to know how exactly that behaviour would be, we might need to pipe the outpt and then take care of intermingling the lines properly? Then we could even add prefixes.

**Comment by syphar** [comment]
Generally, I'm happy to send time on some of the topics here, which do you want to pick? Should I pick?

**Pull request closed without merging**